Adding Online Experimentation data-plane SDK

smichtch opened the pull request:
# Description

Adding Online Experimentation data-plane SDK (for `Microsoft.OnlineExperimentation/workspaces` resources).
TypeSpec PR: https://github.com/Azure/azure-rest-api-specs/pull/33135

# All SDK Contribution checklist:
- [x] **The pull request does not introduce [breaking changes]**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md).**

## [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#developer-guide)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#building-and-unit-testing)
- [x] Pull request includes test coverage for the included changes.

smichtch commented on sdk/onlineexperimentation/azure-analytics-onlineexperimentation/src/main/java/com/azure/analytics/onlineexperimentation/models/ExperimentMetricValidationResult.java:
```diff
@@ -0,0 +1,105 @@
+    @Override
+    public JsonWriter toJson(JsonWriter jsonWriter) throws IOException {
+        jsonWriter.writeStartObject();
+        jsonWriter.writeBooleanField("isValid", this.isValid);
```
Generated code.

smichtch commented on sdk/onlineexperimentation/azure-analytics-onlineexperimentation/src/main/java/com/azure/analytics/onlineexperimentation/models/DiagnosticDetail.java:
```diff
@@ -0,0 +1,96 @@
+    @Generated
+    @Override
+    public JsonWriter toJson(JsonWriter jsonWriter) throws IOException {
+        jsonWriter.writeStartObject();
```
Generated code.

smichtch commented on sdk/onlineexperimentation/azure-analytics-onlineexperimentation/pom.xml:
```diff
@@ -0,0 +1,143 @@
+    </dependency>
+  </dependencies>
+
+  <build>
```
Removed the `<build>` section here, turns out the `<relativePath>` was wrong (had to remove one `../`), works now.

weidongxu-microsoft commented on .vscode/cspell.json:
```diff
@@ -1385,6 +1385,24 @@
+      "filename": "sdk/onlineexperimentation/**",
+      "words": [
+        "momo",
+        "createorupdatemetric",
+        "createorupdatemetricaverage",
+        "createorupdatemetriceventcount",
+        "createorupdatemetriceventrate",
+        "createorupdatemetricpercentile",
+        "createorupdatemetricsum",
+        "createorupdatemetricusercount",
+        "createorupdatemetricuserrate",
+        "vxexdn",
+        "ijchzdh",
```
Some of it comes from generated sample/test.

These files are to help you write your own sample/test. If you write it, you can delete the "generated" folder and get rid of these suppress.
https://github.com/Azure/azure-sdk-for-java/wiki/TypeSpec-Java-Quickstart#samples

weidongxu-microsoft replied:
Yes, you can delete the `generated` folder. It won't cause build problem.

smichtch replied:
@weidongxu-microsoft, I looked through other SDKs, it looks like a good number have keep both generated samples as well as custom made ones. Are you OK to keep this as-is or should I remove them?

smichtch replied:
Cleaned up the generated tests & samples, and reduced cspell.json.

weidongxu-microsoft replied:
Thanks. You can use e.g. `tsp-client update --emitter-optoins="generate-tests=false"`, if you don't want them pop up in next codegen.

You can also do same to turn it off in tspconfig.yaml. But if some of the test/sample helps this time, you may want to keep them there, as in next api-version they might help too.

We likely can have a better sample id in future codegen. Here your JSON `CreateOrUpdateMetric_Average.json` becomes `createorupdatemetricaverage`. Codegen probably better produce `create-or-update-metric-average`.

Yeah, some of existing lib have both co-exist. But that mostly for service not writing enough samples. I see here you already had quite many.

weidongxu-microsoft commented on sdk/onlineexperimentation/azure-analytics-onlineexperimentation/tsp-location.yaml:
```diff
@@ -0,0 +1,4 @@
+directory: specification/onlineexperimentation/Azure.Analytics.OnlineExperimentation
+commit: 7b94be99f4c65d06d2b980843ac54234becbbdda
```
notice, make sure the commit is on specs repo main branch

smichtch replied:
I wanted to keep the commit hash synchronized across all the SDK repos. There haven't been any changes to these specs since that commit, but I can move it forward if this is critical.

weidongxu-microsoft replied:
The requirement is the commit is on "main" branch. The date of the commit is up-to-you. You can have the commit that was 1 year ago here, as long as it's on "main".

Ran a check and seems it's not.

weidongxu-microsoft commented on sdk/onlineexperimentation/azure-analytics-onlineexperimentation/README.md:
```diff
@@ -0,0 +1,78 @@
+
+[Azure Identity][azure_identity] package provides the default implementation for authenticating the client.
+
+## Key concepts
```
Any concept to let user know? Is any resource of high importance?

weidongxu-microsoft commented on sdk/onlineexperimentation/azure-analytics-onlineexperimentation/README.md:
````diff
@@ -0,0 +1,78 @@
+
+## Examples
+
+```java com.azure.analytics.onlineexperimentation.readme
+```
````
Please have at least 1 sample in your README.

You can use any of one in your current samples.

weidongxu-microsoft commented on sdk/onlineexperimentation/azure-analytics-onlineexperimentation/pom.xml:
```diff
@@ -0,0 +1,92 @@
+      <scope>test</scope>
+    </dependency>
+  </dependencies>
+  <profiles>
+    <profile>
+      <id>java21-linux</id>
+      <activation>
+        <jdk>21</jdk>
+        <os>
+          <family>unix</family>
+          <name>linux</name>
+        </os>
+      </activation>
+      <properties>
+        <jacoco.min.branchcoverage>0.25</jacoco.min.branchcoverage>
+      </properties>
+    </profile>
+  </profiles>
```
Why do we need this extra profile?

smichtch replied:
To adjusted the default code coverage threshold, for some reason only linux+jdk21 combination does not meet it.

weidongxu-microsoft reviewed (commented):
basic feedback on non-code part of the PR

We also prefer a comment from EngSys on the split of ci.yml, as confirmation. We need to prepare the pipeline, if the split is confirmed (personally I am fine with it).
https://dev.azure.com/azure-sdk/internal/_build?definitionScope=%5Cjava&pipelineNameFilter=onlineexperimentation

weshaggard commented on sdk/onlineexperimentation/test-resources.bicep:
```diff
@@ -0,0 +1,171 @@
+// PARAMETERS
+@description('The location of the resource. By default, this is the same as the resource group.')
```
Did you forget to create a tests.yml file to run these tests live?

smichtch replied:
Added a tests.yml

weshaggard commented:
> basic feedback on non-code part of the PR
> 
> We also prefer a comment from EngSys on the split of ci.yml, as confirmation. We need to prepare the pipeline, if the split is confirmed (personally I am fine with it). https://dev.azure.com/azure-sdk/internal/_build?definitionScope=%5Cjava&pipelineNameFilter=onlineexperimentation
> 
> Also please resolve the conflict on cspell.

It is fine to split if there is a reason as both are supported however you will need to regenerate the pipelines as mentioned if you keep things split. I also don't see the tests.yml file to go with your test resource scripts.

smichtch commented:
/azp run prepare-pipelines

smichtch commented:
/azp run prepare-pipelines

smichtch commented:
/azp run prepare-pipelines

weidongxu-microsoft commented on sdk/onlineexperimentation/azure-analytics-onlineexperimentation/src/main/java/com/azure/analytics/onlineexperimentation/OnlineExperimentationClient.java:
```diff
@@ -0,0 +1,694 @@
+     * @throws RuntimeException all other wrapped checked exceptions if the request fails to be sent.
+     * @return defines experiment metric metadata and computation details.
+     */
+    public ExperimentMetric activateMetric(String experimentMetricId, String ifMatch) {
```
They seems still be `@ServiceMethod`, as it calls the REST API?

smichtch replied:
Good catch. The async client methods had annotations for the non-generated methods but sync client was missing them.

weidongxu-microsoft reviewed (commented):
A few comments. Overall LGTM.

For 1st release, you also need to get apiview approved (after SDK review meeting).

smichtch commented on sdk/onlineexperimentation/azure-analytics-onlineexperimentation/README.md:
```diff
@@ -0,0 +1,158 @@
+
+### Online Experimentation Workspace
+
+[`Microsoft.OnlineExperimentation/workspaces`][az_exp_workspace] Azure resources work in conjunction with [Azure App Configuration][app_config] and [Azure Monitor][azure_monitor]. The Online Experimentation workspace handles management of metrics definitions and their continuous computation to monitor and evaluate experiment results.
```
Good catch, I'm surprised the linters didn't flag this!

smichtch commented on sdk/onlineexperimentation/azure-analytics-onlineexperimentation/README.md:
```diff
@@ -0,0 +1,158 @@
+}
+
+// [Step 3] Create the experiment metric
+String exampleMetricId = "sample_metric_id_" + (new Random().nextInt(10000) + 10000);
```
Updated to use UUID. Other language SDKs use the random integer method here, copied it here initially.

smichtch commented:
@srnagar please review. Addressed the feedback on APIView: removed non-generated constructors and updated samples/tests accordingly. Added `*WithResponse()` 'overloads' for non-generated client methods.

smichtch commented on sdk/onlineexperimentation/azure-analytics-onlineexperimentation/README.md:
```diff
@@ -0,0 +1,159 @@
+## Troubleshooting
+
+## Next steps
+
```
Updated.

smichtch commented on sdk/onlineexperimentation/test-resources.bicep:
```diff
@@ -0,0 +1,171 @@
+
+var experimentationWorkspaceLocations = [
+  'eastus2'
+  'sweedencentral'
```
Yes a typo, good catch!

smichtch commented on sdk/onlineexperimentation/azure-analytics-onlineexperimentation/assets.json:
```diff
@@ -0,0 +1 @@
+{"AssetsRepo":"Azure/azure-sdk-assets","AssetsRepoPrefixPath":"java","TagPrefix":"java/onlineexperimentation/azure-analytics-onlineexperimentation","Tag": "java/onlineexperimentation/azure-analytics-onlineexperimentation_4b14cfcb73"}
```
Formatted. The file is generated by 'push' command, maybe worth having it format the output.

smichtch commented on sdk/onlineexperimentation/azure-analytics-onlineexperimentation/pom.xml:
```diff
@@ -0,0 +1,93 @@
+        <jacoco.min.linecoverage>0.35</jacoco.min.linecoverage>
+      </properties>
+    </profile>
+  </profiles>
```
Discussed offline, the behavior is unique to JDK 21 on linux. Set the branchcoverage and linecoverage at the project level instead.

smichtch commented on sdk/onlineexperimentation/azure-analytics-onlineexperimentation/src/samples/java/com/azure/analytics/onlineexperimentation/CreateAverageMetricSampleAsync.java:
```diff
@@ -0,0 +1,63 @@
+     */
+    public static void main(String[] args) {
+        createAverageMetricAsync()
+            .block(); // Wait for the operation to complete
```
Updated samples.

smichtch commented on sdk/onlineexperimentation/azure-analytics-onlineexperimentation/ci.yml:
```diff
@@ -0,0 +1,35 @@
+# NOTE: Please refer to https://aka.ms/azsdk/engsys/ci-yaml before editing this file.
+
+trigger:
```
I had a single `sdk/onlineexperimentation/ci.yml` initially and this structure was recommended by the SDK engineering team.

srnagar commented on .vscode/cspell.json:
```diff
@@ -1433,6 +1433,14 @@
+      "words": [
+        "momo",
+        "RESOURCEID",
+        "sweedencentral"
```
nit: This is not required anymore

smichtch commented on sdk/onlineexperimentation/test-resources.bicep:
```diff
@@ -0,0 +1,171 @@
+    }
+  })
+
+  // resource dataExportRule 'dataExports' = {
+  //   name: 'AppEvents'
+  //   properties: {
+  //     destination: {
+  //       resourceId: storageAccount.id
+  //     }
+  //     enable: true
+  //     tableNames: [
+  //       'AppEvents'
+  //       'AppEvents_CL'
+  //     ]
+  //   }
+  // }
```
This was copied from another SDK repo (no way to reference a shared copy, I asked), I'd rather keep the files in sync as-is.
